src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -127,8 +127,18 @@ void primitive_inst::realloc_if_needed() {
 
 void primitive_inst::update_impl() {
     if (!_node.is_type<data>() && !(_node.is_type<mutable_data>() && _node.get_dependencies().empty())) {
-        _impl = std::move(_node.type()->choose_impl(_node));
-        _network.get_program()->compile();
+        auto layout_key = _node.get_output_layout().to_string();
+        auto caches = _node.get_primitive_impl_caches();
+        bool is_new_impl = false;
+        std::tie(_impl, is_new_impl) = caches->get(layout_key, [&]() {
+            cldnn::LRUCache<std::string, std::shared_ptr<cldnn::primitive_impl>>::CacheEntry new_entry;
+            new_entry.data = std::move(_node.type()->choose_impl(_node));
+            new_entry.size = sizeof(new_entry.data);
+            return new_entry;
+        });
+        if (is_new_impl) {
+            _network.get_program()->compile();
```

Comment:
Umm There might be such a case that: 
1) there is a impl in a cache 
2) however that impl is evicted from kernels_cache::in_memory_cache due to the size limit 
3) Then we will be recreate the impl

hmm lets reconsider this flow